File: 20_valid_parentheses.go
func isValid(s string) bool {
    PAIR := map[rune]rune {
        '(': ')',
        '{': '}',
        '[': ']',
    }
    var stack []rune

    for _, str := range s {
        if str == '(' || str == '{' || str == '[' {
            stack = append(stack, str)
            continue
        }
        if len(stack) < 1 {
            return false;
        }
        if str != PAIR[stack[len(stack) - 1]] {
            return false
        }
        stack = stack[:len(stack) - 1]
    }
    return len(stack) == 0

}

File: 20_valid_parentheses.py
PAIR = {
    "(": -1,
    ")": 1,
    "{": -2,
    "}": 2,
    "[": -3,
    "]": 3
}


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for string in s:
            if PAIR[string] < 0:
                stack.append(string)
                continue
            if len(stack) < 1:
                return False
            latest = stack.pop()
            if PAIR[latest] + PAIR[string] != 0:
                return False

        if len(stack) > 0:
            return False

        return True
